build-tools/Xamarin.Android.Tools.BootstrapTasks/Xamarin.Android.Tools.BootstrapTasks/CheckApiCompatibility.cs

```diff
@@ -0,0 +1,227 @@
+				return !Log.HasLoggedErrors;
+			}
+
+			// Get the previous api implementation path by replacing the current api string with the previous one.
+			var previousTargetImplementationPath = TargetImplementationPath.Replace (ApiLevel, previousApiLevel);
+
+			// In case previous api is not defined or directory does not exist we can skip the check.
+			var validateAgainstPreviousApi = !(string.IsNullOrWhiteSpace (previousApiLevel) || !Directory.Exists (previousTargetImplementationPath));
+			if (validateAgainstPreviousApi) {
+
+				// First we check the Api level assembly against the previous api level assembly
+				// i.e.: check api breakages using "the just built V2.dll" against "the just built V1.dll"
+				ValidateApiCompat (previousTargetImplementationPath);
+
+				if (Log.HasLoggedErrors) {
+					return !Log.HasLoggedErrors;
+				}
+			}
+
+			// If Api level is the latest we should also compare it against the reference assembly
+			// located on the external folder. (xamarin-android-api-compatibility)
+			// i.e.: check apicompat using "the just built V2.dll" against V2.dll located on xamarin-android-api-compatibility repo
+			if (ApiLevel == LastStableApiLevel) {
+
+				// Check xamarin-android-api-compatibility reference directory exists
+				var referenceContractPath = Path.Combine (ApiCompatContractPath, "reference");
+				if (!Directory.Exists (referenceContractPath)) {
+					Log.LogMessage (MessageImportance.High, $"CheckApiCompatibility Warning: Skipping reference contract check.\n{referenceContractPath} does not exist.");
+					return !Log.HasLoggedErrors;
+				}
+
+				ValidateApiCompat (referenceContractPath);
+			}
+
+			return !Log.HasLoggedErrors;
+		}
+
+		// Validates Api compatibility between contract (previous version) and implmentation (current version)
+		// We do that by using Microsoft.DotNet.ApiCompat.dll
+		void ValidateApiCompat (string contractPath) {
+
+			var apiCompat = Path.Combine (ApiCompatPath, "Microsoft.DotNet.ApiCompat.exe");
+
+			string content;
+			using (var genApiProcess = new Process()) {
+
+				if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows)){
+  					genApiProcess.StartInfo.FileName = apiCompat;
+					genApiProcess.StartInfo.Arguments = $"{contractPath} -i {TargetImplementationPath}";
+				} else {
+					genApiProcess.StartInfo.FileName = $"mono";
+					genApiProcess.StartInfo.Arguments = $"{apiCompat} {contractPath} -i {TargetImplementationPath}";
+				}
+
+				genApiProcess.StartInfo.UseShellExecute = false;
+				genApiProcess.StartInfo.CreateNoWindow = true;
+				genApiProcess.StartInfo.RedirectStandardOutput = true;
+
+				// Get api definition for previous Api
+				genApiProcess.Start ();
+				content = genApiProcess.StandardOutput.ReadToEnd();
+				genApiProcess.WaitForExit ();
+			}
+
+			ValidateIssues(content);
+		}
+
+		// Validates there is no issue or issues found are acceptable
+		void ValidateIssues(string content) {
+
+			Console.ForegroundColor = ConsoleColor.Red;
```

Comment:
One other thing I see here, is we probably shouldn't use `System.Console` directly from inside an MSBuild task.

They have their own logging methods, so I think you could probably:
* `Log.LogMessage` for general messages
* `Log.LogError` for each error -- and you don't need to turn the console red

You probably won't have to log the total errors either, as MSBuild will print its own summary.

Reply:
The benefit of using MSBuild's logging methods is that things like `.binlog` and custom MSBuild loggers will work. While `System.Console` would only appear in stdout.

Reply:
I tried to use LogError here, but the way Msbuild counts the errors is not the way I want to represent it. It will count one error per line I print, in this case my string has lots of new lines, and I don't want that every single line on my string represent an error.The content of this string contains things that are not errors.

Reply:
You might have to do `LogMessage` for each line, and `LogError` to give the total number of errors if it fails.

But then we'd get those log messages in `binlog` files, too.

Reply:
I update it to use log.Message, but that's very unfortunate because it will not give the same experience I was looking for with Console.WriteLine in Red, but I understand it is important to have those lines on Msbuild log. thanks for point that out.